<a href="https://colab.research.google.com/github/dave502/course.numpy-pandas/blob/main/homework_03/hw6_task_add.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##Дополнительные задания:

Загрузите датасет Wine из встроенных датасетов sklearn.datasets с помощью функции load_wine в переменную data.

In [ ]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_wine
data = load_wine()

Полученный датасет не является датафреймом. Это структура данных, имеющая ключи аналогично словарю. Просмотрите тип данных этой структуры данных и создайте список data_keys, содержащий ее ключи.

In [ ]:
print("type(data) = {}".format(type(data)))
data_keys = data.keys()
data_keys

type(data) = <class 'sklearn.utils.Bunch'>


dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names'])

Просмотрите данные, описание и названия признаков в датасете. Описание нужно вывести в виде привычного, аккуратно оформленного текста, без обозначений переноса строки, но с самими переносами и т.д.

In [ ]:
print(data["DESCR"])

.. _wine_dataset:

Wine recognition dataset
------------------------

**Data Set Characteristics:**

    :Number of Instances: 178 (50 in each of three classes)
    :Number of Attributes: 13 numeric, predictive attributes and the class
    :Attribute Information:
 		- Alcohol
 		- Malic acid
 		- Ash
		- Alcalinity of ash  
 		- Magnesium
		- Total phenols
 		- Flavanoids
 		- Nonflavanoid phenols
 		- Proanthocyanins
		- Color intensity
 		- Hue
 		- OD280/OD315 of diluted wines
 		- Proline

    - class:
            - class_0
            - class_1
            - class_2
		
    :Summary Statistics:
    
    ============================= ==== ===== ======= =====
                                   Min   Max   Mean     SD
    ============================= ==== ===== ======= =====
    Alcohol:                      11.0  14.8    13.0   0.8
    Malic Acid:                   0.74  5.80    2.34  1.12
    Ash:                          1.36  3.23    2.36  0.27
    Alcalinity of Ash:            1

 Данные и названия признаков в датасете

In [ ]:
df = pd.DataFrame(data["data"], columns=data["feature_names"])
df.head()

alcohol  malic_acid   ash  ...   hue  od280/od315_of_diluted_wines  proline
0    14.23        1.71  2.43  ...  1.04                          3.92   1065.0
1    13.20        1.78  2.14  ...  1.05                          3.40   1050.0
2    13.16        2.36  2.67  ...  1.03                          3.17   1185.0
3    14.37        1.95  2.50  ...  0.86                          3.45   1480.0
4    13.24        2.59  2.87  ...  1.04                          2.93    735.0

[5 rows x 13 columns]

Сколько классов содержит целевая переменная датасета? Выведите названия классов.

In [ ]:
print("data['target_names'] = {}".format(data["target_names"]))
print("Количество классов = {}".format(len(data["target_names"])))

data['target_names'] = ['class_0' 'class_1' 'class_2']
Количество классов = 3


На основе данных датасета (они содержатся в двумерном массиве Numpy) и названий признаков создайте датафрейм под названием X.


In [ ]:
X = pd.DataFrame(data["data"], columns=data["feature_names"])
X.head()

alcohol  malic_acid   ash  ...   hue  od280/od315_of_diluted_wines  proline
0    14.23        1.71  2.43  ...  1.04                          3.92   1065.0
1    13.20        1.78  2.14  ...  1.05                          3.40   1050.0
2    13.16        2.36  2.67  ...  1.03                          3.17   1185.0
3    14.37        1.95  2.50  ...  0.86                          3.45   1480.0
4    13.24        2.59  2.87  ...  1.04                          2.93    735.0

[5 rows x 13 columns]

Выясните размер датафрейма X и установите, имеются ли в нем пропущенные значения

In [ ]:
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 178 entries, 0 to 177
Data columns (total 13 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   alcohol                       178 non-null    float64
 1   malic_acid                    178 non-null    float64
 2   ash                           178 non-null    float64
 3   alcalinity_of_ash             178 non-null    float64
 4   magnesium                     178 non-null    float64
 5   total_phenols                 178 non-null    float64
 6   flavanoids                    178 non-null    float64
 7   nonflavanoid_phenols          178 non-null    float64
 8   proanthocyanins               178 non-null    float64
 9   color_intensity               178 non-null    float64
 10  hue                           178 non-null    float64
 11  od280/od315_of_diluted_wines  178 non-null    float64
 12  proline                       178 non-null    float64
dtypes: fl

**В датафрейме 178 объектов, пропусков нет, содержат только числовые значения**

Добавьте в датафрейм поле с классами вин в виде чисел, имеющих тип данных numpy.int64. Название поля - 'target'.

In [ ]:
X["target"] = data["target"]
X["target"].astype(np.int64)
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 178 entries, 0 to 177
Data columns (total 14 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   alcohol                       178 non-null    float64
 1   malic_acid                    178 non-null    float64
 2   ash                           178 non-null    float64
 3   alcalinity_of_ash             178 non-null    float64
 4   magnesium                     178 non-null    float64
 5   total_phenols                 178 non-null    float64
 6   flavanoids                    178 non-null    float64
 7   nonflavanoid_phenols          178 non-null    float64
 8   proanthocyanins               178 non-null    float64
 9   color_intensity               178 non-null    float64
 10  hue                           178 non-null    float64
 11  od280/od315_of_diluted_wines  178 non-null    float64
 12  proline                       178 non-null    float64
 13  targe

Постройте матрицу корреляций для всех полей X. Дайте полученному датафрейму название X_corr.

In [ ]:
X_corr = X.corr()
X_corr

alcohol  malic_acid  ...   proline    target
alcohol                       1.000000    0.094397  ...  0.643720 -0.328222
malic_acid                    0.094397    1.000000  ... -0.192011  0.437776
ash                           0.211545    0.164045  ...  0.223626 -0.049643
alcalinity_of_ash            -0.310235    0.288500  ... -0.440597  0.517859
magnesium                     0.270798   -0.054575  ...  0.393351 -0.209179
total_phenols                 0.289101   -0.335167  ...  0.498115 -0.719163
flavanoids                    0.236815   -0.411007  ...  0.494193 -0.847498
nonflavanoid_phenols         -0.155929    0.292977  ... -0.311385  0.489109
proanthocyanins               0.136698   -0.220746  ...  0.330417 -0.499130
color_intensity               0.546364    0.248985  ...  0.316100  0.265668
hue                          -0.071747   -0.561296  ...  0.236183 -0.617369
od280/od315_of_diluted_wines  0.072343   -0.368710  ...  0.312761 -0.788230
proline                       0.643720   -0.192011  ...  1.000000 -0.633717
target                       -0.328222    0.437776  ... -0.633717  1.000000

[14 rows x 14 columns]

Создайте список high_corr из признаков, корреляция которых с полем target по абсолютному значению превышает 0.5 (причем, само поле target не должно входить в этот список).

In [ ]:
high_corr = [i for i in X_corr.loc[(X_corr["target"]>0.5)&(X_corr["target"]<1), "target"].index]
high_corr

['alcalinity_of_ash']

Удалите из датафрейма X поле с целевой переменной. 


In [ ]:
X.drop("target", axis=1, inplace=True)
X.head(1)

alcohol  malic_acid   ash  ...   hue  od280/od315_of_diluted_wines  proline
0    14.23        1.71  2.43  ...  1.04                          3.92   1065.0

[1 rows x 13 columns]

Для всех признаков, названия которых содержатся в списке high_corr, вычислите квадрат их значений и добавьте в датафрейм X соответствующие поля с суффиксом '_2', добавленного к первоначальному названию признака. Итоговый датафрейм должен содержать все поля, которые, были в нем изначально, а также поля с признаками из списка high_corr, возведенными в квадрат. Выведите описание полей датафрейма X с помощью метода describe.

In [ ]:
X[[s + "_2" for s in high_corr]] = X[high_corr]**2


In [ ]:
X.describe()

alcohol  malic_acid  ...      proline  alcalinity_of_ash_2
count  178.000000  178.000000  ...   178.000000           178.000000
mean    13.000618    2.336348  ...   746.893258           391.142865
std      0.811827    1.117146  ...   314.907474           133.671775
min     11.030000    0.740000  ...   278.000000           112.360000
25%     12.362500    1.602500  ...   500.500000           295.840000
50%     13.050000    1.865000  ...   673.500000           380.250000
75%     13.677500    3.082500  ...   985.000000           462.250000
max     14.830000    5.800000  ...  1680.000000           900.000000

[8 rows x 14 columns]